src/cmap/connection_pool.ts

```diff
@@ -643,6 +650,82 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
     );
   }
 
+  async withConnectionAsync(
+    conn: Connection | undefined,
+    fn: AsyncWithConnectionCallback
+  ): Promise<Connection> {
+    let result;
+    if (conn) {
+      // use the provided connection, and do _not_ check it in after execution
+      try {
+        result = await fn(undefined, conn);
+        return result;
+      } catch (fnErr) {
+        if (fnErr instanceof MongoError && fnErr.code === MONGODB_ERROR_CODES.Reauthenticate) {
+          conn = this.reauthenticateAsync(conn);
+        } else {
+          throw fnErr;
+        }
+      }
+    } else {
+      const { promise, resolve, reject } = promiseWithResolvers<Connection>();
+      this.checkOut((err, conn) => {
+        fn(err as MongoError, conn).then(
+          () => {
+            if (conn) this.checkIn(conn);
+          },
+          fnErr => {
+            if (conn) {
+              this.withReauthenticationAsync(fnErr, conn, fn).then(resolve, reject);
+            } else {
+              reject(fnErr);
+            }
+          }
+        );
+      });
+      return promise;
+    }
+  }
+
+  async withReauthenticationAsync(
+    fnErr: AnyError,
+    conn: Connection,
+    fn: AsyncWithConnectionCallback
+  ): Promise<Connection> {
+    if (fnErr instanceof MongoError && fnErr.code === MONGODB_ERROR_CODES.Reauthenticate) {
+      conn = await this.reauthenticateAsync(conn);
+      return fn(undefined, conn);
+    } else {
+      throw fnErr;
+    }
+  }
```

Comment:
NOTE: These were implemented as a proof of concept to figure out what this logic would look like when converted to async/await before implementing it in Server.command and will be removed before marking as ready for review